// image-js/lib/image/core/environment.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createCanvas = exports.ImageData = exports.DOMImage = void 0;
Object.defineProperty(exports, "createWriteStream", {
  enumerable: true,
  get: function () {
    return _fs.createWriteStream;
  }
});
exports.env = void 0;
exports.fetchBinary = fetchBinary;
Object.defineProperty(exports, "writeFile", {
  enumerable: true,
  get: function () {
    return _fs.writeFile;
  }
});
var _fs = require("fs");
const message = 'requires the canvas library. Install it with `npm install canvas@next`.';
let createCanvas = exports.createCanvas = void 0,
  DOMImage = exports.DOMImage = void 0,
  ImageData = exports.ImageData = void 0;
try {
  // eslint-disable-next-line import/no-unresolved
  const canvas = require('canvas');
  exports.createCanvas = createCanvas = canvas.createCanvas;
  exports.DOMImage = DOMImage = canvas.Image;
  exports.ImageData = ImageData = canvas.ImageData;
} catch (e) {
  exports.createCanvas = createCanvas = function () {
    throw new Error(`createCanvas ${message}`);
  };
  exports.DOMImage = DOMImage = function () {
    throw new Error(`DOMImage ${message}`);
  };
  exports.ImageData = ImageData = function () {
    throw new Error(`ImageData ${message}`);
  };
}
const env = exports.env = 'node';
function fetchBinary(path) {
  return new Promise(function (resolve, reject) {
    (0, _fs.readFile)(path, function (err, data) {
      if (err) reject(err);else resolve(data.buffer);
    });
  });
}
